attest-data/src/messages.rs

```diff
@@ -0,0 +1,620 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::NONCE_SIZE;
+use hubpack::SerializedSize;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+
+use hubpack::error::Error as HubpackError;
+
+/// Magic value for [`Header::magic`]
+pub const ATTEST_MAGIC: u32 = 0xA77E5700;
+
+/// Right now `Attest` is the only command that takes data (nonce)
+pub const MAX_DATA_LEN: usize = NONCE_SIZE;
+
+pub const MAX_REQUEST_SIZE: usize =
+    HostRotHeader::MAX_SIZE + HostToRotCommand::MAX_SIZE + MAX_DATA_LEN;
+
+pub mod version {
+    pub const V1: u32 = 1;
+}
+
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize, SerializedSize,
+)]
+pub struct HostRotHeader {
+    magic: u32,
+    version: u32,
+}
+
+impl Default for HostRotHeader {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl HostRotHeader {
+    pub fn new() -> Self {
+        Self {
+            magic: ATTEST_MAGIC,
+            version: version::V1,
+        }
+    }
+}
+
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize, SerializedSize,
+)]
+#[repr(u32)]
+pub enum HostToRotCommand {
+    /// Returns the certificate chain associated with the RoT
+    GetCertificates,
+    /// Returns the measurement log
+    GetMeasurementLog,
+    /// Calculates sign(sha3_256(hubpack(measurement_log) | nonce))
+    /// and returns the result.
+    Attest,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
+//#[repr(u32)]
+pub enum HostToRotError {
+    _Unused,
+    /// Header magic was incorrect
+    MagicMismatch,
+    /// Mismatch of protocol versions
+    VersionMismatch,
+    /// Message failed to deserialize
+    Deserialize,
+    /// Wrong length of data arguments (expected no data or incorrect length)
+    IncorrectDataLen,
+    /// Unexpected command returned
+    UnexpectedCommand,
+    /// Error return from the sprot command
+    SprotError(SprotError),
```

Comment:
Design question: because we're already supporting nested errors, should we have an error hierarchy here that matches the one in Hubris (variants for `Protocol`, `Spi`, `Attest`, etc) instead of putting them all in `SprotError(..)`?

Reply:
I went back and forth on that. We really only care about errors from `Protocol` `Spi` and `Attest` so it didn't seem that helpful to have a hierarchy vs. a flat error structure.